pkg/server/plugin/keymanager/hashicorpvault/vault_client.go

```diff
@@ -0,0 +1,554 @@
+	}
+
+	for _, keyId := range keyIds {
+		keyIdStr, ok := keyId.(string)
+		if !ok {
+			return nil, status.Errorf(codes.Internal, "expected key id data type %T but got %T", keyIdStr, keyId)
+		}
+
+		keyEntry, err := c.getKeyEntry(ctx, keyIdStr)
+		if err != nil {
+			return nil, err
+		}
+
+		keyEntries = append(keyEntries, keyEntry)
+	}
+
+	return keyEntries, nil
+}
+
+// getKeyEntry gets the transit engine key with the specified spire key id and converts it into a key entry.
+func (c *Client) getKeyEntry(ctx context.Context, spireKeyID string) (*keyEntry, error) {
+	keyData, err := c.getKey(ctx, spireKeyID)
+	if err != nil {
+		return nil, err
+	}
+
+	pk, ok := keyData["public_key"]
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "expected public key to be present")
+	}
+
+	pkStr, ok := pk.(string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "expected public key data type %T but got %T", pkStr, pk)
+	}
+
+	pemBlock, _ := pem.Decode([]byte(pkStr))
+	if pemBlock == nil || pemBlock.Type != "PUBLIC KEY" {
+		return nil, status.Error(codes.Internal, "unable to decode PEM key")
+	}
+
+	pubKeyType, ok := keyData["name"]
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "expected name to be present")
+	}
+
+	pubKeyTypeStr, ok := pubKeyType.(string)
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "expected public key type to be of type %T but got %T", pubKeyTypeStr, pubKeyType)
+	}
+
+	var keyType keymanagerv1.KeyType
+
+	switch pubKeyTypeStr {
+	case "P-256":
+		keyType = keymanagerv1.KeyType_EC_P256
+	case "P-384":
+		keyType = keymanagerv1.KeyType_EC_P384
+	case "rsa-2048":
+		keyType = keymanagerv1.KeyType_RSA_2048
+	case "rsa-4096":
+		keyType = keymanagerv1.KeyType_RSA_4096
+	default:
+		return nil, status.Errorf(codes.Internal, "unsupported key type: %v", pubKeyTypeStr)
+	}
+
+	return &keyEntry{
+		PublicKey: &keymanagerv1.PublicKey{
+			Id:          spireKeyID,
+			Type:        keyType,
+			PkixData:    pemBlock.Bytes,
+			Fingerprint: makeFingerprint(pemBlock.Bytes),
+		},
+	}, nil
+}
+
+// getKey returns a specific key from the transit engine.
+// See: https://developer.hashicorp.com/vault/api-docs/secret/transit#read-key
+func (c *Client) getKey(ctx context.Context, spireKeyID string) (map[string]interface{}, error) {
+	res, err := c.vaultClient.Logical().ReadWithContext(ctx, fmt.Sprintf("/%s/keys/%s", c.clientParams.TransitEnginePath, spireKeyID))
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to get transit engine key: %v", err)
+	}
+
+	keys, ok := res.Data["keys"]
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "transit engine get key call was successful but keys are missing")
+	}
+
+	keyMap, ok := keys.(map[string]interface{})
+	if !ok {
+		return nil, status.Errorf(codes.Internal, "expected key map data type %T but got %T", keyMap, keys)
+	}
+
+	// TODO: Should we support multiple versions of the key?
```

Comment:
I think that SPIRE will always manage a single version of the key, creating a new key when a rotation happens. We need to make sure that keys that have been rotated are being deleted.

Reply:
I see, in that case I will remove this TODO. Thanks for the clarification :slightly_smiling_face: